src/pages/api/revalidate.ts

```diff
@@ -79,37 +96,82 @@ async function handler(request: NextApiRequest, response: NextApiResponse) {
 		)
 	).filter(Boolean) as any[]
 
-	const revalidatePromises = []
+	const pathsToRevalidate = []
 
 	// iterates through nav data recursively and creates a revalidate request
-	function revalidateNavNodes(basePath, nodes) {
+	function constructRevalidateNavNodes(basePath, nodes) {
 		nodes.forEach(({ path, routes }) => {
 			if (routes) {
-				revalidateNavNodes(basePath, routes)
+				constructRevalidateNavNodes(basePath, routes)
 				return
 			}
 
 			if (typeof path !== 'undefined') {
 				const pathToRevalidate = `/${product}/${basePath}/${path}`
-					// remove any trailing slash
-					.replace(/\/$/, '')
 
-				console.log('[revalidate]', pathToRevalidate)
-				revalidatePromises.push(response.revalidate(pathToRevalidate))
+				pathsToRevalidate.push(pathToRevalidate)
 			}
 		})
 	}
 
 	// iterate over each nav node and revalidate its path
 	navDataFiles.forEach(([basePath, navData]) => {
-		revalidateNavNodes(basePath, navData)
+		constructRevalidateNavNodes(basePath, navData)
 	})
 
-	// TODO(brkalow): Add resiliency here, this has the potential to send off hundreds of calls depending on the product, so we should think about how we want to handle network hiccups or partial failure.
-	// wait for everything to get revalidated
-	await Promise.allSettled(revalidatePromises)
+	return pathsToRevalidate
+}
 
-	response.status(200).end()
+/**
+ * Accepts a POST request with a product slug, triggers revalidation for all of a product's docs paths
+ * specified in its latest nav data.
+ */
+async function handler(request: NextApiRequest, response: NextApiResponse) {
+	if (request.method !== 'POST') {
+		response.status(StatusCodes.NOT_FOUND)
+		return
+	}
+
+	// @TODO should we consider a limit on the number of paths?
+	const { product, paths } = request.body
+	const pathsExist = paths || paths.length > 0
+
+	if (!product && !pathsExist) {
+		response
+			.status(StatusCodes.BAD_REQUEST)
+			.json({ error: '[Revalidation failed]: No product or paths provided.' })
+
+		return
+	}
+
+	try {
+		const pathsToRevalidate = pathsExist ? paths : []
+
+		if (product) {
+			const docsPaths = await handleProductDocsRevalidation(product)
+			pathsToRevalidate.push(...docsPaths)
+		}
+
+		const revalidatePromises = []
+
+		pathsToRevalidate.forEach((path: string) => {
+			// remove any trailing slash
+			const formattedPath = path.replace(/\/$/, '')
+			console.log('[revalidate]', formattedPath)
+			revalidatePromises.push(response.revalidate(formattedPath))
+		})
+
+		// TODO(brkalow): Add resiliency here, this has the potential to send off hundreds of calls depending on the product, so we should think about how we want to handle network hiccups or partial failure.
+		// wait for everything to get revalidated
+		await Promise.allSettled(revalidatePromises)
```

Comment:
I want to add support for throttling or batching requests here, perhaps in the separate pr though